### ISSUE TITLE
Interrupting the training procedure of an ML component with a keyboard shortcut

### ISSUE THREAD

**Issue by dakshvar22** [RasaHQ/rasa#7721]
**Description of Problem**: When training a Rasa pipeline on a large dataset, you don't know how many epochs you need to train the ML component for. However, as the training proceeds, monitoring training loss and training accuracy gives an indication of how "saturated" the training is. At that point, I would like to interrupt the training and just save the model checkpoint at the current stage of training. Can we enable that with a keyboard shortcut?

**Overview of the Solution**: A keyboard shortcut to interrupt the training procedure midway. I think it makes sense to do it only if it is lightweight ![](/images/icons/emoticons/biggrin.png)

**Blockers** (if relevant): Not sure how easy it is/best way to implement this.

**Comment by samsucik** [comment]
@joejuzl would you say this can be implemented easily? In case it isn't so straightforward cc @TyDunn because I'm curious to see what you think about this feature.

Also @dakshvar22 I think we could remove this from the research issues board if it's something likely to be owned by engineers, what do you think?

**Comment by joejuzl** [comment]
Are we not already saving the model checkpoint when iterating through the epochs?

If there is anything else we need to do it should not be too hard anyway. In python you can catch `KeyboardInterrupt` which is triggered when the user presses `ctrl-c`, and then run any code you like.
And if it needs to be a specific key press, I doubt it would be more complicated.

**Comment by dakshvar22** [comment]
> Are we not already saving the model checkpoint when iterating through the epochs?

We are but not by default and the use case for that is slightly different. I would use that when I am trying to analyze the model performance across a range of epochs. Whereas, this could be useful when I am training for the first time on a new dataset or training for the first time after making a lot of changes to the dataset and hence do not know what's a good starting number of epochs. Arguably the user could just stick to default number of epochs when training for the first time but they may realize that they don't need so many epochs once the training is started.

As I said, it's only a good-to-have feature and not a necessary feature.

> Also @dakshvar22 I think we could remove this from the research issues board if it's something likely to be owned by engineers, what do you think?

@samsucik I put it in both the boards originally since this might involve playing around with the tensorflow bits as well.

**Comment by samsucik** [comment]
Alright, from what @joejuzl you're saying it sounds like this should be doable by a researcher (at least as a prototype), so I think it makes sense to only keep in the research board. Enable folks can be reached out to by the driving research person when actually needed.

**Comment by koernerfelicia** [comment]
Related: https://github.com/RasaHQ/rasa/issues/6693

**Comment by Ghostvv** [comment]
@dakshvar22 I don't think it is research only issue, since the key input should be communicated to the ML components

**Comment by dakshvar22** [comment]
@Ghostvv What do you mean by this?
>  the key input should be communicated to the ML components

**Comment by Ghostvv** [comment]
the cli should support the key input first, right? I don't think it is research work

**Comment by dakshvar22** [comment]
Do you mean we should give an option to turn on the ability to interrupt through the CLI? I don't think that's needed if we use `ctrl-c` as the standard interruption procedure.
Other than that there isn't anything in the CLI that needs to be done.

**Comment by Ghostvv** [comment]
I see, but then the `ctrl-c` should be propagated to `Agent` to stop training and trigger persistence? I think we'd need help of engineering for that?

**Comment by dakshvar22** [comment]
I was quite curious to try this out and a simple keyboard interrupt works exactly how we want it to work. Pushed the changes to [`interrupt-training` branch](https://github.com/RasaHQ/rasa/compare/interrupt-training). Works well on moodbot. 🤓

**Comment by Ghostvv** [comment]
wow, this is the long waited feature 😍is it possible to stop nlu, and continue training core?

**Comment by dakshvar22** [comment]
Yes! It would just interrupt training of the currently running ML
component(NLU/Core) and then continue with training of subsequent
components in the config.
